README.md
# etch-a-sketch
This is a basic online version of the classic "Etch-A-Sketch" board, in which one can make pixelated sketches on the canvas.

The board supports three possible styles, and two possible modes of applying these styles. The styles can all be combined with the modes to provide six possible variations for sketching.

**Hover** is a mode-type that sketches the canvas by hovering the mouse over it.

**Drag** is a mode-type that sketches the canvas by dragging the mouse over it.

**Default** is a style-type that colors the canvas with solid black.

**Transparent** is a style-type that colors the canvas with low-opacity solid black, which gets progressively darker as each pixel area is passed over repeatedly.

**Colorized** is a style-type that colors the canvas with a random solid color for every pixel area.

The **Clear** button empties the canvas of any applied changes, but retains the selected mode- and style-types.

The **Resize** button creates a prompt that accepts a given value from 1 to 60, which will change the canvas to represent that value for every row and column (for example, passing a value of 20 will create a grid of 400 pixel areas).

Finally, the label **Etch-A-Sketch** also acts as a reset button. Pressing it will remove all applied changes, including on the canvas and the selected mode- and style-types.

That's it!


example-client-server-02/history.txt
   81  cd /b/lib
   82  cd samples/
   83  s
   84  ls
   85  git status .
   86  git pull
   87  git pull -X theirs
   88  mv content content.local
   89  git pull
   90  git pull
   91  ls
   92  git add .
   93  git commit  . -m "add:prototype colorized"
   94  git push
   95  git pull
   96  as
   97  mkdir $(pwd)-downsized
   98  d=$(pwd)-downsized
   99  cp spider__21020921__02_colorized.jpg  $d
  100  cd $d
  101  explorer .
  102  cp ../proto__colorizer__2102112222/spider__21020921__02_colorized.jpg  .
  103  ls
  104  cp spider__21020921__02_colorized.jpg spider__21020921__02_colorized__dns-des.jpg
  105  cd ..
  106  mv proto__colorizer__2102112222-downsized proto__colorizer__2102112222-dns-etc
  107  cd proto__colorizer__2102112222-dns-etc
  108  ls
  109  vi README.md
  110  ls
  111  mv spider__21020921__02_colorized__dns-des.jpg spider__21020921__02_colorized__des-dns.jpg
  112  git add .
  113  git commit . -m < README.md 
  114  git commit . -m "$(cat README.md)" 
  115  git push
  116  git status .
  117  mkdir proto__colorizer__2102112222
  118  cd proto__colorizer__2102112222/
  119  explorer .
  120  scp assignment2.zip  jgi@hu:/data/assignment2.zip
  121  cd
  122  cd Dropbox/lib
  123  cd results/
  124  ls
  125  git pull
  126  ls -ltr
  127  cd ap__210217/
  128  ls
  129  explorer v03
  130  gis-csm ../../v04-1k.cs.jpg
  131   gis-csm ../../v03-1k.cs.jpg
  132  cd /c
  133  ls
  134  cd x
  135  ls
  136  cd ..
  137  rmdir x
  138  ls x__vagrant__200426/
  139  du -ksh x__vagrant__200426/
  140  ls
  141  du -ksh Caishen/
  142  as
  143  as
  144  as
  145  as
  146  as
  147  as
  148  git add contrast-by-color-complement-flowers-field.png 
  149  git add contrast-by-color-complement-flowers-field.png  -f
  150  git pull
  151  git commit . -m add color contralst img"
  152  git push
  153  git commit . -m "add color contralst img"
  154  git push
  155  git push
  156  as
  157  ping as
  158  as
  159  cd
  160  cd kadenze/
  161  ls
  162  ls -l
  163  mkdir ../kadenze.old
  164  mv assignment2/ ../kadenze.old/
  165  ls
  166  ls
  167  cd assignment2/
  168  ls
  169  ls assignment2_1A/
  170  ls assignment2_1A/
  171  ls assignment2_1A/
  172  ls assignment2_1A/
  173  cat assignment2_1A/assignment2_1A.txt
  174  ls assignment2_1B/
  175  cat assignment2_1B/assignment2_1B.txt 
  176  ls
  177  cat assignment2_1B/assignment2_1B.txt 
  178  ls assignment2_1B/
  179  ls assignment2_1B/z
  180  ls assignment2_1B/
  181  ls
  182  ls assignment2_1C/
  183  cat assignment2_1D/assignment2_1D.txt 
  184  cat assignment2_1D/assignment2_1D.txt 
  185  ls assignment2_2/
  186  ls assignment2_2/
  187  ls assignment2_2/
  188  ls
  189  ls -ltr
  190  ls assignment2_1D
  191  ls assignment2_2
  192  ls assignment2_2
  193  ls -ltr
  194  rm assignment2_3B
  195  rm assignment2_3B/*
  196  rm assignment2_3B/*
  197  rm assignment2_3B/*
  198  ls assignment2_3B/
  199  explorer assignment2_3B
  200  cd ..
  201  ls
  202  explorer .
  203  ls -ltr
  204  ls -ltr
  205  cd assignment3/
  206  ls
  207  cd assignment3_1A/
  208  ls -ltr
  209  ls -ltr
  210  ls -ltr
  211  cat assignment3_1A.txt 
  212  cd ..
  213  ls -ltr
  214  ls -ltr assignment3_1B
  215  ls -ltr
  216  ls -ltr assignment3_1C
  217  ls -ltr
  218  ls -ltr assignment3_2
  219  ls -ltr assignment3_2
  220  ls -ltr
  221  ls -ltr
  222  rm -rf assignment3_3B
  223  rm -rf assignment3_3B
  224  rm -rf assignment3_3B
  225  explorer .
  226  ping vpn.nad.ca
  227  cd /c
  228  ls
  229  ls -ltr
  230  ssh jgi@hu
  231  cd
  232  ls .ssh
  233  ls -ltr
  234  cd Dropbox/etc
  235  ls aws/
  236  ls -ltr aws/
  237  cat aws/gaia2_2101080136.pem 
  238  http-server
  239  ssh jgi@hu
  240  git add .
  241  git commit . -m updated
  242  git push
  243  git status .
  244  echo $MSYSTEM
  245  echo $QTDIR
  246  as
  247  as
  248  as
  249  as
  250  as
  251  as
  252  as
  253  as
  254  as
  255  as
  256  as
  257  as
  258  as
  259  as
  260  as
  261  ls
  262  ls -ltr
  263  echo $d
  264  pwd
  265  tar cvzf /c/Users/jeang/Dropbox/Downloads__2101/Backups__Komo_Reader__2103151624.tar.gz .
  266  echo tar cvzf /c/Users/jeang/Dropbox/Downloads__2101/Backups__Komo_Reader__2103151624.tar.gz .
  267  echo tar cvzf /c/Users/jeang/Dropbox/Downloads__2101/Backups__Komo_Reader__2103151624.tar.gz . >>_backup_to_Dropbox__2103151624.sh
  268  cat _backup_to_Dropbox__2103151624.sh 
  269  cp _backup_to_Dropbox__2103151624.sh  /c/Users/jeang/Dropbox/Downloads__2101/Backups__Komo_Reader__2103151624.tar.gz.sh
  270  java -jar Processing_TriggerText_1DTW.jar 
  271  cd ..
  272  ls
  273  java -jar Java/Processing_TriggerText_1DTW.jar 
  274  docker ps
  275  docker commit d2e63217edc8 guillaumeai/runway:picasso
  276  docker ps
  277  docker exec -it d2e63217edc8 bash
  278  cd /c/ma-biblio.wiki/
  279  git pull
  280  cd /c/fichiers/
  281  vi _release.sh 
  282  vi _wiki_publish.sh 
  283  git commit _wiki_publish.sh -m fix:pulling
  284  git push
  285  git pull
  286  git commit _dictations/ -m update
  287  git pull
  288  netstat -nr  | grep 192
  289  cd /c/model/models/
  290  ls
  291  du -ksh *
  292  du -ksh **/*
  293  du -ksh **/checkpoint
  294  rm -rf model_gia-young-picasso-201210_new/checkpoint
  295  rm -rf model_gia-young-picasso-v04-2012211928_new/checkpoint
  296  rm -rf **/checkpoint
  297  du -ksh **/checkpoint
  298  du -ksh **/*
  299  du -ksh **/sample
  300  rm -rf **/sample
  301  du -ksh **/*
  302  du -ksh **/logs;rm -rf **/logs
  303  df -h
  304  vi checkpoint 
  305  netstat -nr
  306  netstat -nr | grep 192
  307  netstat -nr | grep 192
  308  mylocalip
  309  ls /b
  310  ls /b/model/models/
  311  cd /b/model/
  312  git status .
  313  cd models/
  314  ls
  315  ls -ltr
  316  cd /c/model/
  317  git pull
  318  git commit . -m update
  319  git pull
  320  git pull -X theirs
  321  git status .
  322  git commit . -m update:gaia
  323  git merge --abort

script.js
// Variable for grid container element.
const gridcontainer = document.getElementById('gridcontainer');




// Variable for individual box.
const box = document.createElement('div');
box.classList.add('box');
box.style.height = '32px';
box.style.width = '32px';
box.style.opacity = '0.1';
box.style.border = '1px solid gray';

// Variables for buttons.
const headerButton = document.getElementById('header');
const hoverButton = document.getElementById('hover');
const dragButton = document.getElementById('drag');
const clearButton = document.getElementById('clear');
const defaultButton = document.getElementById('default');
const transparentButton = document.getElementById('transparent');
const colorizedButton = document.getElementById('colorized');
const resizeButton = document.getElementById('resize');

// Nodelist for buttons.
const buttons = document.querySelectorAll('button');

// Adds default grid of 16x16 boxes to window on load.
window.onload = function() {
	for (let i = 0; i < 256; i++) {
		gridcontainer.appendChild(box.cloneNode());
	}
}

// Nodelist for boxes.
var boxes = document.getElementsByClassName('box');

// Functions for button stylings.
function changeHeaderButton() {
	headerButton.classList.add('focused');
	setTimeout(() => {buttons.forEach(button => {button.classList.remove('focused')})}, 250);
}

function changeHoverButton() {
	hoverButton.classList.add('focused');
	dragButton.classList.remove('focused');
}

function changeDragButton() {
	dragButton.classList.add('focused');
	hoverButton.classList.remove('focused');
}

function changeDefaultButton() {
	defaultButton.classList.add('focused');
	transparentButton.classList.remove('focused');
	colorizedButton.classList.remove('focused');
}

function changeTransparentButton() {
	transparentButton.classList.add('focused');
	defaultButton.classList.remove('focused');
	colorizedButton.classList.remove('focused');
}

function changeColorizedButton() {
	colorizedButton.classList.add('focused');
	defaultButton.classList.remove('focused');
	transparentButton.classList.remove('focused');
}

function changeClearButton() {
	clearButton.classList.add('focused');
	setTimeout(() => {clearButton.classList.remove('focused')}, 250);
}

// Variable for event handler.
const e = (e) => {};

// Function for default + hover.
function mouseHoverDefault(e) {
	e.target.style.border = 'none';
	e.target.style.opacity = '1';
	e.target.style.background = '#2c2d2d';
}

// Functions for default + drag.
function mouseDragDefaultUp(e) {
 	for (i = 0; i < boxes.length; i++) {
    	boxes[i].removeEventListener('mouseover', mouseHoverDefault);
	} 	
}

function mouseDragDefault(e) {
	e.target.style.border = 'none';
	e.target.style.opacity = '1';
	e.target.style.background = '#2c2d2d';
	for (i = 0; i < boxes.length; i++) {
    	boxes[i].addEventListener('mouseover', mouseHoverDefault);
	}
	for (i = 0; i < boxes.length; i++) {
		boxes[i].addEventListener('mouseup', mouseDragDefaultUp);         					
	}
}

// Function for transparent + hover.
function mouseHoverTransparent(e) {
	e.target.style.border = 'none';
	e.target.style.background = '#2c2d2d'; 
	e.target.style.opacity = String(Number(e.target.style.opacity) + 0.1);
}

// Functions for transparent + drag.
function mouseDragTransparentUp(e) {
	e.target.style.border = 'none';
	e.target.style.background = '#2c2d2d';		
 	for (i = 0; i < boxes.length; i++) {
    	boxes[i].removeEventListener('mouseover', mouseHoverTransparent);
	} 
}

function mouseDragTransparent(e) {
	e.target.style.border = 'none';
	e.target.style.background = '#2c2d2d'; 
	e.target.style.opacity = String(Number(e.target.style.opacity) + 0.1);
	for (i = 0; i < boxes.length; i++) {
    	boxes[i].addEventListener('mouseover', mouseHoverTransparent);
	}
	for (i = 0; i < boxes.length; i++) {
		boxes[i].addEventListener('mouseup', mouseDragTransparentUp);        					
	}
}

// Function for colorized + hover.
function mouseHoverColorized(e) {
	const hue = 'rgb(' + (Math.floor(Math.random() * 256)) + ',' + (Math.floor(Math.random() * 256)) + ',' + (Math.floor(Math.random() * 256)) + ')';
	e.target.style.border = 'none';
	e.target.style.background = hue;
	e.target.style.opacity = '0.5';
}

// Functions for colorized + drag.
function mouseDragColorizedUp(e) {
	const hue = 'rgb(' + (Math.floor(Math.random() * 256)) + ',' + (Math.floor(Math.random() * 256)) + ',' + (Math.floor(Math.random() * 256)) + ')';
	e.target.style.border = 'none';

	e.target.style.background = hue; 
	e.target.style.opacity = '0.5';
 	for (i = 0; i < boxes.length; i++) {
    	boxes[i].removeEventListener('mouseover', mouseHoverColorized);
    }
}

function mouseDragColorized(e) {
	const hue = 'rgb(' + (Math.floor(Math.random() * 256)) + ',' + (Math.floor(Math.random() * 256)) + ',' + (Math.floor(Math.random() * 256)) + ')';
	e.target.style.border = 'none';
	e.target.style.background = hue;
	e.target.style.opacity = '0.5';
	for (i = 0; i < boxes.length; i++) {
    	boxes[i].addEventListener('mouseover', mouseHoverColorized);
	}
	for (i = 0; i < boxes.length; i++) {
		boxes[i].addEventListener('mouseup', mouseDragColorizedUp);
    }
}

// Functions for all mode combinations.
function defaultHover() {
	for (let i = 0; i < boxes.length; i++) {
		boxes[i].addEventListener('mouseenter', mouseHoverDefault);
	}
}

function defaultDrag() {
	for (let i = 0; i < boxes.length; i++) {	
   		boxes[i].addEventListener('mousedown', mouseDragDefault);
    }
}

function transparentHover() {
	for (let i = 0; i < boxes.length; i++) {
		boxes[i].addEventListener('mouseenter', mouseHoverTransparent);
	}
}

function transparentDrag() {
	for (let i = 0; i < boxes.length; i++) {	
   		boxes[i].addEventListener('mousedown', mouseDragTransparent);
   	}
}

function colorizedHover() {
	for (let i = 0; i < boxes.length; i++) {
		boxes[i].addEventListener('mouseenter', mouseHoverColorized);
	}
}

function colorizedDrag() {
	for (let i = 0; i < boxes.length; i++) {	
   		boxes[i].addEventListener('mousedown', mouseDragColorized);
   	}
}

// Functions for deselecting other modes.
function removeHoverDefault() {
	for (let i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mouseenter', mouseHoverDefault);
	}
}

function removeDragDefaultUp() {
	for (i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mouseup', mouseDragDefaultUp);         					
	}
}

function removeDragDefault() {
	for (let i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mousedown', mouseDragDefault);
	}	
}

function removeHoverTransparent() {
	for (i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mouseenter', mouseHoverTransparent);
	}
}

function removeDragTransparentUp() {
	for (i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mouseup', mouseDragTransparentUp);
	}
}

function removeDragTransparent() {
	for (i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mousedown', mouseDragTransparent);
	}
}

function removeHoverColorized() {
	for (i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mouseenter', mouseHoverColorized);
	}
}

function removeDragColorizedUp() {
	for (i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mouseup', mouseDragColorizedUp);
	}
}

function removeDragColorized() {
	for (i = 0; i < boxes.length; i++) {
		boxes[i].removeEventListener('mousedown', mouseDragColorized);
	}
}

// Checks what buttons have been selected, then chooses the corresponding function.
function checkSelected() {
	const selectedButtons = Array.from(document.querySelectorAll('.focused'));

	// Code to run for hover + default.
	if (selectedButtons.includes(document.querySelector('button#hover.focused')) && selectedButtons.includes(document.querySelector('button#default.focused'))) {
		gridcontainer.style.cursor = 'pointer';

		removeHoverTransparent();
		removeHoverColorized();
		removeDragDefault();
		removeDragDefaultUp();
		removeDragTransparent();
		removeDragTransparentUp();
		removeDragColorized();
		removeDragColorizedUp();

		defaultHover();
	}

	// Code to run for drag + default.
	else if (selectedButtons.includes(document.querySelector('button#drag.focused')) && selectedButtons.includes(document.querySelector('button#default.focused'))) {
		gridcontainer.style.cursor = 'pointer';

		removeDragTransparent();
		removeDragTransparentUp();
		removeDragColorized();
		removeDragColorizedUp();
		removeHoverDefault();
		removeHoverTransparent();
		removeHoverColorized();

		defaultDrag();
	}

		// Code to run for hover + transparent.
	else if (selectedButtons.includes(document.querySelector('button#hover.focused')) && selectedButtons.includes(document.querySelector('button#transparent.focused'))) {
		gridcontainer.style.cursor = 'pointer';

		removeHoverDefault();
		removeHoverColorized();
		removeDragDefault();
		removeDragDefaultUp();
		removeDragTransparent();
		removeDragTransparentUp();
		removeDragColorized();
		removeDragColorizedUp();

		transparentHover();
	}

		// Code to run for drag + transparent.
	else if (selectedButtons.includes(document.querySelector('button#drag.focused')) && selectedButtons.includes(document.querySelector('button#transparent.focused'))) {
		gridcontainer.style.cursor = 'pointer';

		removeDragDefault();
		removeDragDefaultUp();
		removeDragColorized();
		removeDragColorizedUp();
		removeHoverDefault();
		removeHoverTransparent();
		removeHoverColorized();

		transparentDrag();
	}

		// Code to run for hover + colorized.
	else if (selectedButtons.includes(document.querySelector('button#hover.focused')) && selectedButtons.includes(document.querySelector('button#colorized.focused'))) {
		gridcontainer.style.cursor = 'pointer';

		removeHoverDefault();
		removeHoverTransparent();
		removeDragDefault();
		removeDragDefaultUp();
		removeDragTransparent();
		removeDragTransparentUp();
		removeDragColorized();
		removeDragColorizedUp();

		colorizedHover();
	}

		// Code to run for drag + colorized.
	else if (selectedButtons.includes(document.querySelector('button#drag.focused')) && selectedButtons.includes(document.querySelector('button#colorized.focused'))) {
		gridcontainer.style.cursor = 'pointer';

		removeDragDefault();
		removeDragDefaultUp();
		removeDragTransparent()
		removeDragTransparentUp();
		removeHoverDefault();
		removeHoverTransparent();
		removeHoverColorized(); 

		colorizedDrag();
	}

	else {
		return;
	}
}

// Function for clearing grid.
function clearGrid() {
	for (let i = 0; i < boxes.length; i++) {
		boxes[i].style.opacity = '0.1';
		boxes[i].style.border = '1px solid black';
		boxes[i].style.background = '#f9f9f9';
	}
}

//Function for resizing grid.
var resizeNumber;

function resizeGrid() {
	resizeButton.classList.add('focused');

	setTimeout(() => {
		resizeNumber = String(parseInt(prompt('Please enter how many grids and columns you would like. The maximum number allowed is 60.')));
		var boxesArray = Array.from(boxes);
		try {			
			etchResized();
		} catch (error) {
			console.log("Tried to ran : etchResized();" + error);
		}

		if (resizeNumber <= 60 && resizeNumber > 0) {
			for (let i = 0; i < boxesArray.length; i++) {
				gridcontainer.removeChild(document.querySelector('.box'));
			}

			var newBox = document.createElement('div');
			newBox.classList.add('box');
			newBox.style.height = String(512 / Number(resizeNumber)) + 'px';
			newBox.style.width = String(512 / Number(resizeNumber)) + 'px';
			newBox.style.opacity = '0.1';
			newBox.style.border = '1px solid black';

			for (let i = 0; i < (resizeNumber * resizeNumber); i++) {
				gridcontainer.appendChild(newBox.cloneNode());
			}

			checkSelected();		
		}

		else if (resizeNumber <= 0 || resizeNumber > 60) {
			return;
		}

		else if (resizeNumber === null || resizeNumber === undefined) {
			return;
		}
	}, 250);

	setTimeout(() => {resizeButton.classList.remove('focused')}, 250);
}

// Function for resetting entire grid.
function resetAll() {
	clearGrid();
	checkSelected();

	removeDragDefault();
	removeDragDefaultUp();

	removeDragTransparent()
	removeDragTransparentUp();

	removeDragColorized();
	removeDragColorizedUp();

	removeHoverDefault();
	removeHoverTransparent();
	removeHoverColorized(); 

	var boxesArray = Array.from(boxes);

	for (let i = 0; i < boxesArray.length; i++) {
		gridcontainer.removeChild(document.querySelector('.box'));
	}

	for (let i = 0; i < 256; i++) {
		gridcontainer.appendChild(box.cloneNode());
	}

	gridcontainer.style.cursor = 'default';
}

// Event listeners for changing button style.
headerButton.addEventListener('click', changeHeaderButton);
hoverButton.addEventListener('click', changeHoverButton);
dragButton.addEventListener('click', changeDragButton);
clearButton.addEventListener('click', changeClearButton);
defaultButton.addEventListener('click', changeDefaultButton);
transparentButton.addEventListener('click', changeTransparentButton);
colorizedButton.addEventListener('click', changeColorizedButton);

// Event listeners for selecting button modes.
headerButton.addEventListener('click', resetAll);
hoverButton.addEventListener('click', checkSelected);
dragButton.addEventListener('click', checkSelected);
clearButton.addEventListener('click', clearGrid);
defaultButton.addEventListener('click', checkSelected);
transparentButton.addEventListener('click', checkSelected);
colorizedButton.addEventListener('click', checkSelected);
resizeButton.addEventListener('click', resizeGrid);

